heksher/setting.py

```diff
@@ -80,47 +87,42 @@ def get(self, **contexts) -> T:
             try:
                 from_rules = self.last_ruleset.resolve(contexts, self)
             except NoMatchError:
-                from_rules = MISSING
+                from_rules = NO_DEFAULT
         else:
             logger.warning('the value of setting was never retrieved from service', extra={'setting': self.name})
-            from_rules = MISSING
+            from_rules = NO_DEFAULT
 
-        if from_rules is MISSING:
-            if self.default_value is MISSING:
+        if from_rules is NO_DEFAULT:
+            if self.default_value is NO_DEFAULT:
                 raise NoMatchError(self.name)
             return self.default_value
         return from_rules
 
-    def update(self, client, context_features: Sequence[str], rules: Iterable[Tuple[Sequence[Tuple[str, str]], T]],
-               is_rule_collection: bool = True):
+    def update(self, client, context_features: Sequence[str], rules: Iterable[Tuple[Sequence[Tuple[str, str]], T]]):
         """
         Update a setting's rules from a client
 
         Args:
             client: The client updating the setting.
             context_features: The context features the root was collated by.
             rules: An iterable of rules to collate.
-            is_rule_collection: bool, if rules are a collection; if not, don't collate the rules
         """
         if self.last_ruleset:
             last_client = self.last_ruleset.client()
             if last_client and last_client is not client:
                 logger.warning('setting received rule set from multiple clients',
                                extra={'setting': self.name, 'new_client': client, 'last_client': last_client})
-        if is_rule_collection:
-            updated_rules: List[Tuple[Sequence[Tuple[str, str]], T]] = []
-            for rule in rules:
-                rule_ = rule[0]
-                value_ = rule[1]
-                for callback in self._callbacks:
-                    value_ = callback(self, rule_, value_)  # type: ignore
-                updated_rules.append((rule_, value_))
+
+        def validate(rule: Sequence[Tuple[str, str]], value: T) -> T:
+            for validator in self._validators:
+                value = validator(value, rule, self)
+            return value
+
+        if context_features:
+            updated_rules = [(rule, validate(rule, value)) for (rule, value) in rules]
             root = collate_rules(context_features, updated_rules)
         else:
-            root_ = rules
-            for callback in self._callbacks:
-                root_ = callback(self, '', root_)  # type: ignore
-            root = root_  # type: ignore
+            root = validate((), rules)  # type: ignore[arg-type]
```

Comment:
unnecessary, if there are no contexts features you should still collate a rule set, it's the user's responsibility to call update with valid data